src/main/java/software/amazon/cloudformation/Wrapper.java

```diff
@@ -0,0 +1,567 @@
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.exc.MismatchedInputException;
+import com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+import software.amazon.awssdk.awscore.exception.AwsServiceException;
+import software.amazon.awssdk.http.SdkHttpClient;
+import software.amazon.awssdk.http.apache.ApacheHttpClient;
+import software.amazon.awssdk.utils.StringUtils;
+import software.amazon.cloudformation.exceptions.BaseHandlerException;
+import software.amazon.cloudformation.exceptions.FileScrubberException;
+import software.amazon.cloudformation.exceptions.TerminalException;
+import software.amazon.cloudformation.injection.CloudWatchLogsProvider;
+import software.amazon.cloudformation.injection.CloudWatchProvider;
+import software.amazon.cloudformation.injection.CredentialsProvider;
+import software.amazon.cloudformation.injection.SessionCredentialsProvider;
+import software.amazon.cloudformation.loggers.CloudWatchLogHelper;
+import software.amazon.cloudformation.loggers.CloudWatchLogPublisher;
+import software.amazon.cloudformation.loggers.LogPublisher;
+import software.amazon.cloudformation.metrics.MetricsPublisher;
+import software.amazon.cloudformation.metrics.MetricsPublisherImpl;
+import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
+import software.amazon.cloudformation.proxy.Credentials;
+import software.amazon.cloudformation.proxy.DelayFactory;
+import software.amazon.cloudformation.proxy.HandlerErrorCode;
+import software.amazon.cloudformation.proxy.HandlerRequest;
+import software.amazon.cloudformation.proxy.LoggerProxy;
+import software.amazon.cloudformation.proxy.MetricsPublisherProxy;
+import software.amazon.cloudformation.proxy.OperationStatus;
+import software.amazon.cloudformation.proxy.ProgressEvent;
+import software.amazon.cloudformation.proxy.ResourceHandlerRequest;
+import software.amazon.cloudformation.proxy.WaitStrategy;
+import software.amazon.cloudformation.resource.ResourceTypeSchema;
+import software.amazon.cloudformation.resource.SchemaValidator;
+import software.amazon.cloudformation.resource.Serializer;
+import software.amazon.cloudformation.resource.Validator;
+import software.amazon.cloudformation.resource.exceptions.ValidationException;
+
+public abstract class Wrapper<ResourceT, CallbackT> {
```

Comment:
Could we name it as `AbstractWrapper` to avoid misuse it.

Reply:
That is a much better name. Will use that instead